rust/gitxetcore/src/stream/stdout_process_stream.rs

```diff
@@ -0,0 +1,285 @@
+use parutils::AsyncIterator;
+
+use super::data_iterators::AsyncDataIterator;
+
+/// An async data tterator that sources the data from the stdout of a process running in the background.
+///
+/// Example usage smudging git lfs pointer.
+///
+///    let mut command = tokio::process::Command::new("git");
+///    command.current_dir(lfs_repo_dir).arg("lfs").arg("smudge");
+///
+///    // Streams out smudged data.
+///    let data_out = AsyncStdoutDataIterator::from_command(command, &lfs_pointer_file[..], 16*1024*1024)
+///
+pub struct AsyncStdoutDataIterator {
+    child_process: Option<tokio::process::Child>,
+    stdout: tokio::process::ChildStdout,
+    bufsize: usize,
+}
+
+impl AsyncStdoutDataIterator {
+    pub async fn from_command(
+        mut command: tokio::process::Command,
+        stdin_data: &[u8],
+        stdout_bufsize: usize,
+    ) -> Result<Self> {
+        let mut process = command
+            .stdin(Stdio::piped())
+            .stdout(Stdio::piped())
+            .stderr(Stdio::piped()) // This one to be captured.
+            .spawn()?;
+
+        // Get the stdin handle and write the data
+        let Some(mut stdin) = process.stdin.take() else {
+            Err(anyhow!(
+                "Unable to connect stdin in child process {command:?}.",
+            ))?;
+            unreachable!();
+        };
+
+        stdin.write_all(stdin_data).await?; // Send all data
+        stdin.flush().await?; // Ensure all data is sent
+
+        Self::from_process(process, stdout_bufsize)
+    }
+
+    /// Constructs an Async Data Iterator reader that pipes the output from the stdout
+    /// of a process into a data iterator.
+    /// Reads things into an internal
+    /// buffer of a given size.
+    /// It is not guaranteed that every read will
+    /// fill the complete buffer.
+    pub fn from_process(mut child_process: tokio::process::Child, bufsize: usize) -> Result<Self> {
+        // Read the command's output from stdout in a streaming manner
+        let Some(stdout) = child_process.stdout.take() else {
+            Err(anyhow!("Unable to connect stdout in child process."))?;
+            unreachable!();
+        };
+
+        Ok(Self {
+            child_process: Some(child_process),
+            stdout,
+            bufsize,
+        })
+    }
+}
+
+#[async_trait]
+impl AsyncIterator<GitXetRepoError> for AsyncStdoutDataIterator {
+    type Item = Vec<u8>;
+
+    async fn next(&mut self) -> Result<Option<Self::Item>> {
+        let mut buffer = vec![0u8; self.bufsize];
+
+        let readlen = self.stdout.read(&mut buffer).await?;
+
+        if readlen > 0 {
+            buffer.resize(readlen, 0);
+            Ok(Some(buffer))
+        } else {
+            let Some(child) = self.child_process.take() else {
+                Err(anyhow!("Next called after child process completed."))?;
+                unreachable!();
+            };
+
+            let output = child.wait_with_output().await?;
```

Comment:
I smell a bit danger here. `read` returning 0 does not mean that the reader will always no longer be able to produce bytes so there may be remaining bytes in this output.

Reply:
Good catch.  Updated this and read up on all the cases where we should do a retry.  Rust's io read api is dumb.